freeglut/freeglut/src/mswin/fg_spaceball_mswin.c

```diff
@@ -42,158 +50,563 @@
+#define A3DX_EVENT_BUTTON     0x03	/* bitwise buttons (all devices) */
+#define A3DX_EVENT_BUTTON_BDF 0x16	/* bitwise dual function (SM Pro's) */
+#define V3DX_EVENT_BUTTON_X   0x1c	/* integer key codes (SM Enterprise) */
+#define V3DX_EVENT_BUTTON_XDF 0x1d	/* integer dual function (SM Enterprise) */
+
+#define A3DX_DEFAULT_BUTTON_BITS 16	/* default valid bitwise button bits */
+#define V3DX_MAX_KEYS 0xff			/* max value of integer key codes */
+
+uint16_t fg_sball_buttons = 0;		/* number of buttons */
+uint16_t fg_sball_bitwise_bits = 0;	/* bitwise event 0x03 data size in bits*/
+
+/* Multi-axis Controller is 0x01, 0x08 and RIDEV_DEVNOTIFY is arrival/removal */
+RAWINPUTDEVICE multiAxis = { 0x01, 0x08, RIDEV_DEVNOTIFY, 0x00 };
+
+struct SFG_SpaceballConfig {
+    uint16_t VID;
+    uint16_t PID;
+    uint8_t bits;		/* device specific valid button data bits */
+    uint8_t buttons;	/* number of buttons (physical + virtual) */
 };
+typedef struct SFG_SpaceballConfig SFG_SpaceballConfig;
 
-extern int fg_sball_initialized;
-unsigned int __fgSpaceKeystate = 0;
-RAWINPUTDEVICE __fgSpaceball = { 0x01, 0x08, 0x00, 0x00 };
+/* Spaceball device list with: VID, PID, bits and buttons */
+static SFG_SpaceballConfig spaceballList[] = {
+    {0x046d, 0xc603, 16, 10},   /* spacemouse plus XT - different HID schema? */
+    {0x046d, 0xc605, 16, 4},    /* cadman */
+    {0x046d, 0xc606, 16, 8},    /* spacemouse classic */
+    {0x046d, 0xc621, 16, 12},   /* spaceball 5000 */
+    {0x046d, 0xc623, 16, 8},    /* space traveller */
+    {0x046d, 0xc625, 24, 21},   /* space pilot */
+    {0x046d, 0xc626, 16, 2},    /* space navigator */
+    {0x046d, 0xc627, 16, 15},   /* space explorer */
+    {0x046d, 0xc628, 16, 2},    /* space navigator for notebooks */
+    {0x046d, 0xc629, 32, 31},   /* space pilot pro */
+    {0x046d, 0xc62b, 32, 19},   /* space mouse pro, 15+4 */
+    {0x046d, 0xc640, 16, 4},    /* nulooq            - different HID schema? */
+    {0x256f, 0xc62e, 16, 2},    /* spacemouse wireless (USB cable) 15+4 */
+    {0x256f, 0xc62f, 16, 2},    /* spacemouse wireless  receiver,  15+4 */
+    {0x256f, 0xc631, 32, 19},   /* spacemouse pro wireless */
+    {0x256f, 0xc632, 32, 19},   /* spacemouse pro wireless receiver */
+    {0x256f, 0xc633, 32, 51},   /* spacemouse enterprise 31+20 */
+    {0x256f, 0xc635, 16, 2},    /* spacemouse compact */
+    {0x256f, 0xc636, 16, 2},    /* spacemouse module */
+    {0x256f, 0xC641, 16, 2},    /* spacemouse module (other) */
+
+    {0, 0, 0, 0}
+};
+
+/* sets the fg_sball_bitwise_bits and fg_sball_buttons based on VID and PID */
+void fgSpaceballUpdateDeviceList(void);
 
 void fgPlatformInitializeSpaceball(void)
 {
-    HWND hwnd;
-    fg_sball_initialized = 1;
-    if (!fgStructure.CurrentWindow)
+    if( fg_sball_initialized == 1 )
+        return;
+
+    if( !fgStructure.CurrentWindow )
     {
-        fg_sball_initialized = 0;
+        fg_sball_initialized = -1;
```

Comment:
Here you changed the initialization of fg_sball_initialized to -1, but further on all tests are `if(!fg_sball_initialized)`, am I missing something?

Reply:
nope, you caught something! Changing back to = 0.